djstripe/models.py

```diff
@@ -676,74 +497,31 @@ def extend(self, delta):
         self.customer.sync_current_subscription()
 
 
-class Invoice(StripeObject):
-    stripe_api_name = "Invoice"
+class Invoice(StripeInvoice):
+    objects = models.Manager()
 
     customer = models.ForeignKey(Customer, related_name="invoices")
-    attempted = models.NullBooleanField()
-    attempts = models.PositiveIntegerField(null=True)
-    closed = models.BooleanField(default=False)
-    paid = models.BooleanField(default=False)
-    period_end = models.DateTimeField()
-    period_start = models.DateTimeField()
-    subtotal = models.DecimalField(decimal_places=2, max_digits=7)
-    total = models.DecimalField(decimal_places=2, max_digits=7)
-    date = models.DateTimeField()
-    charge = models.CharField(max_length=50, blank=True)
 
     class Meta:
         ordering = ["-date"]
 
-    def __str__(self):
-        return "<total={total}, paid={paid}, stripe_id={stripe_id}>".format(total=self.total, paid=smart_text(self.paid), stripe_id=self.stripe_id)
-
-    def retry(self):
-        if not self.paid and not self.closed:
-            inv = self.api_retrieve()
-            inv.pay()
-            return True
-        return False
-
-    def status(self):
-        if self.paid:
-            return "Paid"
-        if self.closed:
-            return "Closed"
-        return "Open"
-
     @classmethod
     def sync_from_stripe_data(cls, stripe_invoice, send_receipt=True):
         c = Customer.objects.get(stripe_id=stripe_invoice["customer"])
-        period_end = convert_tstamp(stripe_invoice, "period_end")
-        period_start = convert_tstamp(stripe_invoice, "period_start")
-        date = convert_tstamp(stripe_invoice, "date")
-
-        invoice, created = cls.objects.get_or_create(
-            stripe_id=stripe_invoice["id"],
-            defaults=dict(
-                customer=c,
-                attempted=stripe_invoice["attempted"],
-                closed=stripe_invoice["closed"],
-                paid=stripe_invoice["paid"],
-                period_end=period_end,
-                period_start=period_start,
-                subtotal=stripe_invoice["subtotal"] / decimal.Decimal("100"),
-                total=stripe_invoice["total"] / decimal.Decimal("100"),
-                date=date,
-                charge=stripe_invoice.get("charge") or ""
-            )
-        )
+
+        try:
+            invoice = cls.stripe_objects.get_by_json(stripe_invoice)
+            created = False
+        except cls.DoesNotExist:
+            invoice = cls.create_from_stripe_object(stripe_invoice)
+            invoice.customer = c
+            created = True
```

Comment:
I've done so many things down stream from this branch :). It sounds great but I'd rather save that idea for the 162 branch